### PR TITLE
Fix bytes.replace

### PR DESCRIPTION
**Pull request by youknowone**

### CODE REVIEW

File: vm/src/obj/objbyteinner.rs [hunk 1]
```diff
@@ -1096,41 +1096,165 @@ impl PyByteInner {
...
+
+    fn replace_delete(&self, from: PyByteInner, maxcount: Option<usize>) -> Vec<u8> {
+        let count = count_substring(self.elements.as_slice(), from.elements.as_slice(), maxcount);
+        if count == 0 {
+            // no matches
+            return self.elements.clone();
+        }
+
+        let result_len = self.len() - (count * from.len());
+        debug_assert!(self.len() >= count * from.len());
```
Comment (palaviv) [inline vm/src/obj/objbyteinner.rs:1123]: maybe remove this? I know is is only in debug but it does not provide additional data.

File: vm/src/obj/objbyteinner.rs [hunk 2]
```diff
@@ -1096,41 +1096,165 @@ impl PyByteInner {
...
+        to: PyByteInner,
+        maxcount: Option<usize>,
+    ) -> Vec<u8> {
+        let len = from.len();
+        let mut iter = self.elements.find_iter(&from.elements);
+
+        let mut new = if let Some(offset) = iter.next() {
+            let mut new = self.elements.clone();
+            new[offset..offset + len].clone_from_slice(to.elements.as_slice());
+            if maxcount == Some(1) {
```
Comment (palaviv) [inline vm/src/obj/objbyteinner.rs:1152]: The logic here is a little confusing. Maybe we should do the rest of the function in the `else` of this?

File: vm/src/obj/objbyteinner.rs [hunk 3]
```diff
@@ -1096,41 +1096,165 @@ impl PyByteInner {
...
+    ) -> PyResult<Vec<u8>> {
+        let count = count_substring(self.elements.as_slice(), from.elements.as_slice(), maxcount);
+        if count == 0 {
+            // no matches, return unchanged
+            return Ok(self.elements.clone());
+        }
+
+        // Check for overflow
+        //    result_len = self_len + count * (to_len-from_len)
+        debug_assert!(count > 0);
```
Comment (palaviv) [inline vm/src/obj/objbyteinner.rs:1186]: Do we need this?

File: vm/src/obj/objbyteinner.rs [hunk 4]
```diff
@@ -1202,6 +1326,23 @@ pub fn try_as_byte(obj: &PyObjectRef) -> Option<Vec<u8>> {
     })
 }
 
+#[inline]
+fn count_substring(haystack: &[u8], needle: &[u8], maxcount: Option<usize>) -> usize {
+    let substrings = haystack.find_iter(needle);
+    if let Some(maxcount) = maxcount {
+        let mut count = 0;
```
Comment (palaviv) [inline vm/src/obj/objbyteinner.rs:1333]: How about `min(substring.count(), maxcount)`?

File: vm/src/obj/objbyteinner.rs (hunk 1 above)
Comment (youknowone) [reply vm/src/obj/objbyteinner.rs:1123]: This is line by line port of CPython implementation. Do you still want to remove this regardless of it?

File: vm/src/obj/objbyteinner.rs (hunk 4 above)
Comment (youknowone) [reply vm/src/obj/objbyteinner.rs:1333]: When `substring.count()` is big and `maxcount` is small, the iteration cost will be different by early break.

Comment (youknowone) [reply vm/src/obj/objbyteinner.rs]: Oh, I used it with `take()` and it works. Thanks!

File: vm/src/obj/objbyteinner.rs (hunk 1 above)
Comment (palaviv) [reply vm/src/obj/objbyteinner.rs:1154]: I am OK with you keeping this. Just not sure if that is necessary. Your call.

**Pull request merged**